### PR TITLE
rgw: adjust the request_uri to support absoluteURI of http request

### PR DESCRIPTION
**Pull request by vmcloud**
http://tracker.ceph.com/issues/12917 Fixes: #12917

Signed-off-by: Wenjun Huang <wenjunhuang@tencent.com>

### CODE REVIEW

**Comment by dachary** [comment]
@vmcloud I'm not able to review the content of the pull request. Just a comment on the commit message: it needs to be **Fixes: #12917** instead of **Fixes: 12917**

**Comment by vmcloud** [comment]
Hi, @dachary 
Sorry for my carelessness before, I have update the commit messages and added some comments in the related source code. Thanks!

**Comment by dachary** [comment]
@vmcloud thanks :-)

File: src/rgw/rgw_common.cc [hunk 1]
```diff
@@ -97,11 +97,29 @@ is_err() const
...
 }
 
+// The requestURI transferred from the frontend can be abs_path or absoluteURI
+// If it is absoluteURI, we should adjust it to abs_path for the following 
+// S3 authorization and some other processes depending on the requestURI
+// The absoluteURI can start with "http://", "https://", "ws://" or "wss://"
+static string get_abs_path(const string &request_uri) {
+  int beg_pos = request_uri.find("://") + 3;
+  int len = request_uri.size();
+  while(beg_pos < len) {
```
Comment (free-luowei) [inline src/rgw/rgw_common.cc]: why not use find here?

Comment (vmcloud) [inline src/rgw/rgw_common.cc]: Yes, it is a good suggestion.

**Comment by liewegas** [comment]
@cbodley @yehudasa @mattbenjamin needs review?

File: src/rgw/rgw_common.cc [hunk 2]
```diff
@@ -97,11 +97,48 @@ is_err() const
...
+  }
+  return true;
+}
+
+// The requestURI transferred from the frontend can be abs_path or absoluteURI
+// If it is absoluteURI, we should adjust it to abs_path for the following 
+// S3 authorization and some other processes depending on the requestURI
+// The absoluteURI can start with "http://", "https://", "ws://" or "wss://"
+static string get_abs_path(const string& request_uri) {
+  const string ABS_PREFIXS[] = {"http://", "https://", "ws://", "wss://"};
```
Comment (cbodley) [inline src/rgw/rgw_common.cc]: can we make this array static so we don't construct the std::strings on every call?

Comment (vmcloud) [inline src/rgw/rgw_common.cc]: Yes, good advice!

**Comment by vmcloud** [comment]
Have made the const array static

**Comment by tchaikov** [comment]
@vmcloud did you removed your remote branch of wip-12917 by accident? as a side effect, it closes the pull request automatically.

**Comment by vmcloud** [comment]
@tchaikov yes, I am trying to reopen it.

**Comment by vmcloud** [comment]
I have recreate a PR for this bug: https://github.com/ceph/ceph/pull/7675

**Pull request closed without merging**